spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/bigquant/SpatialConvolution.scala

```diff
@@ -0,0 +1,382 @@
+import com.intel.analytics.bigdl.nn.abstractnn.{Initializable, TensorModule}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor.{FloatType, QuantTensor, Tensor}
+import com.intel.analytics.bigdl.utils.serializer.{DataConverter, ModuleData}
+import com.intel.analytics.bigdl.utils.{T, Table}
+import java.io.{IOException, ObjectInputStream, ObjectOutputStream}
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe
+import serialization.Bigdl.{AttrValue, BigDLModule}
+
+@SerialVersionUID(- 8008252944905538960L)
+class SpatialConvolution[T: ClassTag](
+  val nInputPlane: Int, // The number of expected input planes in the image given into forward()
+  val nOutputPlane: Int, // The number of output planes the convolution layer will produce.
+  val kernelW: Int, // The kernel width of the convolution
+  val kernelH: Int, // The kernel height of the convolution
+  val strideW: Int = 1, // The step of the convolution in the width dimension.
+  val strideH: Int = 1, // The step of the convolution in the height dimension
+  val padW: Int = 0, // The additional zeros added per width to the input planes.
+  val padH: Int = 0, // The additional zeros added per height to the input planes.
+  val nGroup: Int = 1 // Kernel group number
+)(implicit ev: TensorNumeric[T]) extends QuantModule[T](nOutputPlane) with Initializable {
+
+  require(nInputPlane % nGroup == 0, "Number of input channels should be multiples of group.")
+  require(nOutputPlane % nGroup == 0, "Number of output channels should be multiples of group.")
+
+  @transient var weight: Array[QuantTensor[T]] = null
+  var data: QuantTensor[T] = QuantTensor[T]()
+
+  val bias: Tensor[T] = Tensor[T](nOutputPlane)
+
+  val FAULT_TOLERANCE = 0.5f
+  val WEIGHT_THRESHOLD = 64.0f
+  val THRESHOLD = 127.0f
+  val DILATION_HEIGHT = 1
+  val DILATION_WIDTH = 1
+
+  @transient var _init = false
+
+  private def setWeightSum(weight: Tensor[T], weightSum: Array[T], group: Int): Unit = {
+    val start = nOutputPlane / nGroup * (group - 1)
+    for (i <- 1 to weight.size(1)) {
+      val singleRow = weight.select(1, i)
+      weightSum(start + i - 1) = singleRow.sum()
+    }
+  }
+
+  private def outputSize(input: Int, pad: Int, kernel: Int, stride: Int): Int = {
+    (input + 2 * pad - kernel) / stride + 1
+  }
+
+
+  def initWeightAndBias(weightFP32: Tensor[T], biasFP32: Tensor[T]): this.type = {
+    if (biasFP32 != null) {
+      bias.copy(biasFP32)
+    } else {
+      bias.fill(ev.fromType(0)) // TODO bias may be null, at that time, we should not initialize it
+    }
+
+    // Reshape the weight. Computing the min, max, and quantizing the weight don't need group
+    for (group <- 1 to nGroup) {
+      val weightFP32Tmp = weightFP32.select(1, group)
+      setWeightSum(weightFP32Tmp, weightSum, group)
+    }
+
+    val weightFP32Tmp = weightFP32.view(Array(nOutputPlane, nInputPlane / nGroup, kernelH, kernelW))
```

Comment:
Only NCWH is supported, yes?

Reply:
NHWC also be supported. This feature will be added soon.